lib/ts/recipe/emailpassword/api/implementation.ts

```diff
@@ -620,26 +623,78 @@ export default function getAPIImplementation(): APIInterface {
+
+            const mfaInstance = MultiFactorAuthRecipe.getInstance();
+
+            // we do not want to attempt accountlinking when there is an active session and MFA is turned on
+            if (session === undefined) {
+                // the above sign in recipe function does not do account linking - so we do it here.
+                response.user = await AccountLinking.getInstance().createPrimaryUserIdOrLinkAccounts({
+                    tenantId,
+                    user: response.user,
+                    userContext,
+                });
+            }
+
+            if (mfaInstance === undefined) {
+                // No MFA stuff here, so we just create and return the session
+                let session = await Session.createNewOrKeepExistingSession(
+                    options.req,
+                    options.res,
+                    tenantId,
+                    emailPasswordRecipeUser.recipeUserId,
+                    {},
+                    {},
+                    userContext
+                );
+
+                return {
+                    status: "OK",
+                    session,
+                    user: response.user,
+                };
+            }
+
+            const mfaValidationRes = await mfaInstance.validateForMultifactorAuthBeforeFactorCompletion({
+                req: options.req,
+                res: options.res,
                 tenantId,
-                user: response.user,
+                factorIdInProgress: "emailpassword",
+                session,
+                userLoggingIn: response.user,
+                isAlreadySetup: true,
                 userContext,
             });
 
-            let session = await Session.createNewSession(
-                options.req,
-                options.res,
+            if (mfaValidationRes.status !== "OK") {
+                return mfaValidationRes;
+            }
+
+            const sessionRes = await mfaInstance.createOrUpdateSessionForMultifactorAuthAfterFactorCompletion({
+                req: options.req,
+                res: options.res,
                 tenantId,
-                emailPasswordRecipeUser.recipeUserId,
-                {},
-                {},
-                userContext
-            );
+                factorIdInProgress: "emailpassword",
+                justCompletedFactorUserInfo: {
+                    user: response.user,
+                    createdNewUser: false,
+                    recipeUserId: emailPasswordRecipeUser.recipeUserId,
+                },
+                userContext,
+            });
+
+            if (sessionRes.status !== "OK") {
+                return sessionRes;
+            }
+
             return {
                 status: "OK",
-                session,
-                user: response.user,
+                session: sessionRes.session,
+                user: (await getUser(response.user.id, userContext))!, // fetching user again cause the user might have been updated while setting up mfa
```

Comment:
we should not use `!` and handle race condition properly (in case it's undefined, and session exists, we want to throw unauthorised error), else if session doesn't exist, then we can throw a regular JS error

Reply:
Done